### PR TITLE
Pull out ptr metadata operations into an "emap" module

### PR DESCRIPTION
**Pull request by davidtgoldblatt**
This pulls out the associations between addresses, edatas, szind, and slab status into its own module (mostly pulling from the extent and rtree code).

The immediate motivation is that much of the splitting, merging, and boundary modification code can now be reused in an upcoming hugepage-focused allocator. There are a few ancillary benefits:
- Extent merge / splitting logic is now simpler to understand and better document
- Metadata lookups are less syntactically cumbersome (rtree_ctx setup is moved into helpers).
- We squeeze out a few performance optimizations in rtree lookup code.

After this diff stack, the emap is a strict wrapper of the rtree that adds a locking strategy and some convenience accessors. The latter is an implementation detail of the former, and no code outside the emap module uses any rtree-specific knowledge.

### CODE REVIEW

File: src/extent.c [hunk 1]
```diff
@@ -357,27 +357,6 @@ extent_reregister(tsdn_t *tsdn, edata_t *edata) {
 	assert(!err);
 }
 
-/*
- * Removes all pointers to the given extent from the global rtree indices for
- * its interior.  This is relevant for slab extents, for which we need to do
- * metadata lookups at places other than the head of the extent.  We deregister
- * on the interior, then, when an extent moves from being an active slab to an
- * inactive state.
- */
```
[the remark this answers is not in the document]
Comment (davidtgoldblatt) [reply src/extent.c]: I thought this was well-covered by the new documentation in emap.h. The boundary/interior split is described  on emap_register_interior, and, that given, the deregister variants seem straightforward. Will add back if you still disagree.

File: include/jemalloc/internal/emap.h [hunk 2]
```diff
@@ -79,7 +79,12 @@ struct emap_prepare_s {
  * is going to be used as a slab, you still need to call emap_register_interior
  * on it, though.
  *
- * Each operation has a "prepare" and a "commit" portion.  The prepare portion
+ * Remap simply changes the szind and slab status of an extent's boundary
+ * mappings.  If the extent is not a slab, it doesn't bother with updating the
```
[the remark this answers is not in the document]
Comment (davidtgoldblatt) [reply include/jemalloc/internal/emap.h]: Right; the end is part of the boundary. But this is only called as part of activation and deactivation, and in both cases we know we're protected from concurrent merge attempts. I'll clarify the comment.

File: src/extent.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (interwq) [reply src/extent.c]: No I left the comment when I only saw the removal.

**Comment by davidtgoldblatt** [comment]
FreeBSD failures are the CI packaging issue we've been seeing recently -- merging.

**Pull request merged**